packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/FishjamClientInternal.kt

```diff
@@ -1,175 +1,729 @@
+    val websocketListener =
+      object : WebSocketListener() {
+        override fun onClosed(
+          webSocket: WebSocket,
+          code: Int,
+          reason: String
+        ) {
+          listener.onSocketClose(code, reason)
+        }
+
+        override fun onMessage(
+          webSocket: WebSocket,
+          bytes: ByteString
+        ) {
+          try {
+            val peerMessage = PeerNotifications.PeerMessage.parseFrom(bytes.toByteArray())
+            if (peerMessage.hasAuthenticated()) {
+              listener.onAuthSuccess()
+              commandsQueue.finishCommand()
+            } else if (peerMessage.hasMediaEvent()) {
+              receiveEvent(peerMessage.mediaEvent.data)
+            } else {
+              Timber.w("Received unexpected websocket message: $peerMessage")
+            }
+          } catch (e: Exception) {
+            Timber.e("Received invalid websocket message", e)
           }
+        }
+
+        override fun onOpen(
+          webSocket: WebSocket,
+          response: Response
+        ) {
+          listener.onSocketOpen()
+          val authRequest =
+            PeerNotifications.PeerMessage
+              .newBuilder()
+              .setAuthRequest(
+                PeerNotifications.PeerMessage.AuthRequest
+                  .newBuilder()
+                  .setToken(config.token)
+              ).build()
+          sendEvent(authRequest)
+        }
+
+        override fun onFailure(
+          webSocket: WebSocket,
+          t: Throwable,
+          response: Response?
+        ) {
+          listener.onSocketError(t, response)
+        }
+      }
+
+    commandsQueue.addCommand(
+      Command(CommandName.CONNECT, ClientState.CONNECTED) {
+        val request = Request.Builder().url(config.websocketUrl).build()
+        val webSocket =
+          OkHttpClient().newWebSocket(
+            request,
+            websocketListener
+          )
+
+        this.webSocket = webSocket
+      }
+    )
+  }
+
+  fun join(peerMetadata: Metadata = emptyMap()) {
+    commandsQueue.addCommand(
+      Command(CommandName.JOIN, ClientState.JOINED) {
+        localEndpoint = localEndpoint.copy(metadata = peerMetadata)
+        rtcEngineCommunication.connect(peerMetadata)
+      }
+    )
+  }
+
+  override fun onConnected(
+    endpointID: String,
+    otherEndpoints: List<com.fishjamcloud.client.events.Endpoint>
+  ) {
+    localEndpoint = localEndpoint.copy(id = endpointID)
+
+    otherEndpoints.forEach {
+      var endpoint = Endpoint(it.id, EndpointType.fromString(it.type), it.metadata)
+
+      for ((trackId, trackData) in it.tracks) {
+        val track = Track(null, it.id, trackId, trackData.metadata ?: mapOf())
+        endpoint = endpoint.addOrReplaceTrack(track)
+
+        this.listener.onTrackAdded(track)
+      }
+      this.remoteEndpoints[it.id] = (endpoint)
+    }
+    listener.onJoined(endpointID, remoteEndpoints)
+    commandsQueue.finishCommand()
```

Comment:
What ensures that first command is Connected?
If i am not mistaken sb can call join on empty queue, then first command is `Join`, not `Connected`

Reply:
That somebody is us, in FishjamClientInternal. User just calls `connect`, we are calling connect and then join.